Better behaviour in the presence of 429s

j-baker opened the pull request:
If a client sends requests at a rate of 40 per second, but the server
demands only 20 per second by sending back 429s, the default behaviour
should be that the server should smooth the client's load down to 20
per second.

Right now this does not work, because we use exponential backoff with
a maximum number of retries, and each call is independent, so each
individual request is as likely to succeed as any other.

What this means is that given sustained attempted load of 40 per second
and a limit of 20 requests per second, eventually some of the calls
_will_ fail.

You can see this most easily right now with an endpoint that takes 1ms
to return with a rate limit of 1 request per second, which will fail
within a few seconds.

This PR attempts to fix the issue. It uses Netflix's concurrency-limits
library to apply a tcp-like strategy to request concurrency; basically
provided a single request per second won't be limited into oblivion,
you can make progress.

I also include a test that will fail immediately unless this concurrency
limiting code exists.

At the moment, the scope is per-endpoint - I don't know how to do
better. We could later on define 'limiting domains' which get somehow
annotated by the service author according to rate limit groupings.

Note that we'd have to modify our internal retrofit codegen to properly
add the path template header.

There is scope to improve this later - particularly to do this on the
server too, and prioritise interactive over batch traffic.

The main downside of this code is that you have to make sure you always
clean up - otherwise you have a resource leak. In order to avoid this,
there is code which cleans up in the case of the resource leak, logging
at the same time. This should avoid the IOException deadlock pain
we've seen in the past.

j-baker committed:
Better behaviour in the presence of backing off

I assert that:

If a client sends requests at a rate of 40 per second, but the server
demands only 20 per second by sending back 429s, the default behaviour
should be that the server should smooth the client's load down to 20
per second.

Right now this does not work, because we use exponential backoff with
a maximum number of retries, and each call is independent, so each
individual request is as likely to succeed as any other.

What this means is that given sustained attempted load of 40 per second
and a limit of 20 requests per second, eventually some of the calls
_will_ fail.

You can see this most easily right now with an endpoint that takes 1ms
to return with a rate limit of 1 request per second, which will fail
within a few seconds.

This PR attempts to fix the issue. It uses Netflix's concurrency-limits
library to apply a tcp-like strategy to request concurrency; basically
provided a single request per second won't be limited into oblivion,
you can make progress.

I also include a test that will fail immediately unless this concurrency
limiting code exists.

At the moment, the scope is per-endpoint - I don't know how to do
better. We could later on define 'limiting domains' which get somehow
annotated by the service author according to rate limit groupings.

Note that we'd have to modify our internal retrofit codegen to properly
add the path template header.

There is scope to improve this later - particularly to do this on the
server too, and prioritise interactive over batch traffic.

The main downside of this code is that you have to make sure you always
clean up - otherwise you have a resource leak. In order to avoid this,
there is code which cleans up in the case of the resource leak, logging
at the same time. This should avoid the IOException deadlock pain
we've seen in the past.

j-baker commented:
Note - testing is insufficient right now - this is to get validation on the approach, though I would like to get this merged soon :)

j-baker committed:
Merge branch 'develop' into jbaker/better_429_behaviour

j-baker commented on http-clients/src/main/java/com/palantir/remoting3/clients/ClientConfigurations.java:
```diff
@@ -87,7 +87,7 @@ public static ClientConfiguration of(
                 .proxy(ProxySelector.getDefault())
                 .proxyCredentials(Optional.empty())
-                .maxNumRetries(uris.size())
+                .maxNumRetries(DEFAULT_MAX_NUM_RETRIES)
```
@samrogerson fysa

uschi2000 replied:
uhm, I thought we had merged such a change already?

j-baker replied:
yes, but we did it in only one of the two places (see above in this class)

uschi2000 commented on okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/ConcurrencyLimiters.java:
```diff
@@ -0,0 +1,147 @@
+
+    public AsyncLimiter limiter(Request request) {
+        final String limiterKey;
+        String pathTemplate = request.header(OkhttpTraceInterceptor.PATH_TEMPLATE_HEADER);
```
this is a bit dodgy

uschi2000 commented on okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/ConcurrencyLimiters.java:
```diff
@@ -0,0 +1,147 @@
+        return limiter(limiterKey);
+    }
+
+    static final class AsyncLimiter {
```
this construction is pretty complicated. minimally needs some docs to explain what's going on.

uschi2000 commented on okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/ConcurrencyLimiters.java:
```diff
@@ -0,0 +1,147 @@
+        private final Queue<SettableFuture<Limiter.Listener>> waitingRequests = new LinkedBlockingQueue<>();
+        private final Limiter<Void> limiter;
+
+        public AsyncLimiter(
```
why do we need to make the limiter asynchronous? from https://github.com/Netflix/concurrency-limits/blob/master/concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/Limiter.java it seems like Limiter#acquire returns immediately?

uschi2000 replied:
oh i see, nvm.

uschi2000 commented on okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/ConcurrencyLimiters.java:
```diff
@@ -0,0 +1,147 @@
+                if (head == null) {
+                    acquired.onIgnore();
+                } else {
+                    head.set(acquired);
```
does this construction satisfy some basic fairness properties, i.e., every request will get scheduled/acquired eventually?

j-baker replied:
yes, via the FIFO queue

uschi2000 commented on okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/ConcurrencyLimiters.java:
```diff
@@ -0,0 +1,147 @@
+                    return;
+                }
+                Limiter.Listener acquired = maybeAcquired.get();
+                SettableFuture<Limiter.Listener> head = waitingRequests.poll();
```
would prefer to make the multi-threadedness here a little easier to understand by calling `poll` only
```
SettableFuture<> head;
// Note that different threads may be executing processQueue; this is safe because ...
while( (head = waitingRequests.poll()) != null ) {
  ..
}
```

j-baker replied:
ah but if you do that, you're not guaranteed to have any permits to give them

uschi2000 replied:
i see. then add a comment, please

j-baker replied:
done

uschi2000 commented on okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/RemotingOkHttpCall.java:
```diff
@@ -166,14 +194,18 @@ public void onFailure(Call call, IOException exception) {
                 // Relay successful responses
                 if (response.code() / 100 <= 2) {
                     callback.onResponse(call, response);
+                    listener.onSuccess();
```
the construction here is very brittle because you need to trace that the Listener is released (onSuccess, onDropped) on all code paths.

j-baker replied:
yeah, i'm gonna clean that up

uschi2000 commented on okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/RemotingOkHttpCall.java:
```diff
@@ -236,14 +272,18 @@ public Void visit(QosException.Throttle exception) {
                 }
 
                 Duration backoff = exception.getRetryAfter().orElse(nonAdvertizedBackoff.get());
```
difficult to grok what the interplay is between the Limiter and the backoffStrategy. they seem duplicative.

j-baker replied:
and yet, they are not :)

uschi2000 replied:
exactly

j-baker commented on okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/OkHttpClients.java:
```diff
@@ -221,7 +221,8 @@ private static RemotingOkHttpClient createInternal(
                 client.build(),
-                () -> new ExponentialBackoff(config.maxNumRetries(), config.backoffSlotSize(), new Random()),
+                () -> new ExponentialBackoff(
+                        config.maxNumRetries(), config.backoffSlotSize(), ThreadLocalRandom.current()),
```
I use ThreadLocalRandom here because Random will cause requests that start in the same millisecond to back off in the exact same way, which is undesirable.

markelliot replied:
if you want those guarantees it seems marginally safer to have ExponentialBackoff take a `Supplier<Random>` instead of a concrete `Random` (admittedly I haven't yet traced the RemotingOkHttpClient code to check that it makes an ExponentialBackoff per thread)

j-baker replied:
it makes an ExponentialBackoff per call

j-baker replied:
basically it just needs something stable and random - threadlocalrandom is perfect here.

j-baker committed:
more docs, easier to read

uschi2000 commented on okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/ConcurrencyLimiters.java:
```diff
@@ -0,0 +1,177 @@
+ * limitations under the License.
+ */
+
+package com.palantir.remoting3.okhttp;
```
@iamdanfox what's the deal with remoting-vs-conjure in PRs?

uschi2000 commented on okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/ConcurrencyLimiters.java:
```diff
@@ -0,0 +1,177 @@
+        }
+
+        private Limiter.Listener wrap(
+                Map<Limiter.Listener, Runnable> activeListeners, Limiter.Listener listener) {
```
remove activeListeners param?

uschi2000 replied:
maybe even make ConcurrencyLimiter non-static?

j-baker replied:
done

uschi2000 commented on okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/ConcurrencyLimiters.java:
```diff
@@ -0,0 +1,177 @@
+     * Upon a request finishing, we check if there are any waiting requests, and if there are we attempt to trigger
+     * some more.
+     */
+    static final class ConcurrencyLimiter {
```
private?

j-baker replied:
nah, this is used

uschi2000 commented on okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/ConcurrencyLimiters.java:
```diff
@@ -0,0 +1,177 @@
+
+    private final ConcurrentMap<String, ConcurrencyLimiter> limiters = new ConcurrentHashMap<>();
+
+    private static Limiter<Void> newLimiter() {
```
inline in ConcurrencyLimiter constructor?

j-baker replied:
done

uschi2000 commented on okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/ConcurrencyLimiters.java:
```diff
@@ -0,0 +1,177 @@
+
+    ConcurrencyLimiter limiter(Request request) {
+        final String limiterKey;
+        String pathTemplate = request.header(OkhttpTraceInterceptor.PATH_TEMPLATE_HEADER);
```
this is still dodgy

j-baker replied:
I don't _really_ see a way of avoiding this. It seems reasonable to do this by endpoint, and if you do that you end up with this.

ellisjoe replied:
could also see this being something that uses a dynamic proxy which makes it much easier to limit per method or per some annotation. think the only sad thing about this is relying on the tracing header which is only every passed around internally (never sent across the wire)

uschi2000 replied:
OK. But then I'd rename the code bits so that they're no longer "trace"-specific. Probably also need to stop deleting the header in the trace-specific code path

uschi2000 commented on okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/RemotingOkHttpCall.java:
```diff
@@ -185,81 +211,125 @@ public void onResponse(Call call, Response response) throws IOException {
-                Optional<Duration> nonAdvertizedBackoff = backoffStrategy.nextBackoff();
-                if (!nonAdvertizedBackoff.isPresent()) {
-                    callback.onFailure(call, new IOException("Failed to reschedule call since "
+                Optional<Duration> nonAdvertisedBackoff = backoffStrategy.nextBackoff();
```
thanks

uschi2000 commented on okhttp-clients/src/test/java/com/palantir/remoting3/okhttp/ExponentialBackoffTest.java:
```diff
@@ -54,4 +74,63 @@ public void testRetriesCorrectNumberOfTimesAndFindsRandomBackoffWithInExponentia
+            public void run() {
+                for (int i = 0; i < 1001;) {
+                    Limiter.Listener listener = Futures.getUnchecked(limiter.acquire());
+                    //System.out.println(i);
```
meep

j-baker replied:
I'm going to rewrite these tests into a kinda contract test thing.

uschi2000 commented on okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/RemotingOkHttpCall.java:
```diff
@@ -140,6 +149,23 @@ private static ResponseBody buffer(MediaType mediaType, byte[] bodyBytes) {
+            public void onFailure(Throwable t) {
+                callback.onFailure(
+                        RemotingOkHttpCall.this,
+                        new IOException(new AssertionError("This should never happen", t)));
```
explain a bit more in the message please

j-baker replied:
done

uschi2000 commented on okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/ConcurrencyLimiters.java:
```diff
@@ -0,0 +1,177 @@
+
+/**
+ * Remoting calls may observe 429 or 503 responses from the server, at which point they back off in order to
+ * reduce excess load. Unfortunately this state on backing off is stored per-call, so 429s or 503s in one call do not
```
maybe just describe how it works rather lamenting ("Unfortunately") the decomp?
```
Flow control in Conjure is a collaborative effort between servers and clients: Servers advertise an overloaded state via 429/503 responses, and clients throttle the number of requests they are sending. The latter is implemented as a combination of two techniques, yielding a mechanism similar to flow control in TCP/IP: First, clients use the frequency of 429/503 responses to determine an estimate for the number of permissible concurrent requests. Second, each such request gets scheduled according to an exponential backoff algorithm. This class provides an asynchronous implementation of Netflix's concurrency-limits library for determining the above-mentioned concurrency estimates. [...]
```

j-baker replied:
done, thanks

j-baker committed:
Better concurrency limiters

j-baker committed:
fixes

uschi2000 reviewed (approved):
I'm fine with the functionality. Decomp-wise, I sort of wonder if we should roll the limit and backoff functionality into the same class? Don't know exactly what that would look like.

Waiting for @iamdanfox 's feedback to see where we should merge this.

j-baker commented:
got fixes to this comin in soon

j-baker commented:
and tests etc

j-baker committed:
simplify

ellisjoe commented on okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/ConcurrencyLimiters.java:
```diff
@@ -0,0 +1,144 @@
+final class ConcurrencyLimiters {
+    private static final String FALLBACK = "";
+
+    private final ConcurrentMap<String, ConcurrencyLimiter> limiters = new ConcurrentHashMap<>();
```
can we make this a size limited cache?

j-baker committed:
Checkstyle

ellisjoe commented on okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/ConcurrencyLimiters.java:
```diff
@@ -0,0 +1,144 @@
+    }
+
+    ConcurrencyLimiter limiter(Request request) {
+        final String limiterKey;
```
I'm always pro moving out logic into a function when you have this conditional initialization. Could also just make this `String limiterKey = limiter != null ? blah : FALLBACK;`

j-baker replied:
done

ellisjoe commented on okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/ConcurrencyLimiters.java:
```diff
@@ -0,0 +1,144 @@
+
+        private synchronized void processQueue() {
+            while (!waitingRequests.isEmpty()) {
+                Optional<Limiter.Listener> maybeAcquired = limiter.acquire(null);
```
what does `null` represent? would making it a named constant make this more readable?

j-baker replied:
renamed to NO_CONTEXT

ellisjoe commented on okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/RemotingOkHttpCall.java:
```diff
@@ -185,81 +212,125 @@ public void onResponse(Call call, Response response) throws IOException {
+
+        private void retry(Call retryCall, Duration delay) {
+            scheduleExecution(() -> {
+                listener.onDropped();
```
why does this `retry` call `onDropped` but the one above calls `onIgnore`?

ellisjoe replied:
also, doesn't it make sense to call this outside of the scheduled execution? that's the point at which you know the server was overloaded and doing marking the dropped request here would skew the timings

j-baker replied:
For posterity: When the server tells you to back off for a certain time, it's conceptually saying 'I can service your request at this point, please don't send me any more requests until then'. So it's reasonable to delay giving back permits to run other concurrent requests requests until that point.

ellisjoe commented on okhttp-clients/src/test/java/com/palantir/remoting3/okhttp/FlowControlTest.java:
```diff
@@ -0,0 +1,156 @@
+ * This class is a simulation of the flow control primitives used by this library, in order to allow the developer
+ * to try different strategies.
+ * <p>
+ * It is run in CI, but only to prevent code breakages - this is in general a
```
"...in general a"?

ellisjoe commented on okhttp-clients/src/test/java/com/palantir/remoting3/okhttp/FlowControlTest.java:
```diff
@@ -0,0 +1,156 @@
+        ListenableFuture<?> task = Futures.allAsList(tasks);
+        while (!task.isDone()) {
+            sleep(1000);
+            log.info("Average rate is {}, 1 minute rate is {}", rate.getMeanRate(), rate.getOneMinuteRate());
```
`assertThat(rate.getMeanRate()).isBetween(19, 21)`? not sure where these values are expected to be but some sort of broad assertion seems reasonable.

j-baker replied:
It's actually kinda interesting, you don't expect to necessarily get that close to the maximum.

Basically if the target concurrency rate is small enough, it won't be able to compensate right. So for example in this case, you basically cycle between:

1 thread, 10 requests a second.
2 threads, 20 requests a second, perfect
3 threads, 30 requests a second, fails
1 thread, 10 requests a second
2 thread, 20 requests a second
etc etc

and you basically end up at about 16, overall.

ellisjoe commented on okhttp-clients/src/test/java/com/palantir/remoting3/okhttp/FlowControlTest.java:
```diff
@@ -0,0 +1,156 @@
+    @Test
+    public void test16ThreadsRateLimit20() throws ExecutionException, InterruptedException {
+        Meter rate = new Meter();
+        List<ListenableFuture<?>> tasks = createWorkers(rate, 16, 20, Duration.ofMillis(100))
```
would also be interesting to see what happens when you add more workers

j-baker replied:
turns out the number of workers doesn't actually affect this until the rate limit is also large (because waiting isn't factored into this calculation).

ellisjoe commented on okhttp-clients/src/test/java/com/palantir/remoting3/okhttp/FlowControlTest.java:
```diff
@@ -0,0 +1,156 @@
+            }
+        }
+
+        private void initializeBackoff() {
```
just use `Suppliers.memoize`

j-baker replied:
I do clear this variable. Are you sure you want this?

ellisjoe commented on okhttp-clients/src/test/java/com/palantir/remoting3/okhttp/ConcurrencyLimitersTest.java:
```diff
@@ -0,0 +1,58 @@
+import org.junit.Test;
+
+public final class ConcurrencyLimitersTest {
+    private final ConcurrencyLimiters limiters = new ConcurrencyLimiters();
```
injecting the initial limit here would have made the second test easier to understand

j-baker replied:
done

j-baker committed:
checkstyle

ellisjoe commented on okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/ConcurrencyLimiters.java:
```diff
@@ -0,0 +1,144 @@
+            return future;
+        }
+
+        private synchronized void processQueue() {
```
personally had a hard time understanding how this works/when this gets called given it can be called either when you `acquire` or when you notify the listener that you're done.

j-baker replied:
Chatted to joe in person, he's now ok with this I think.

ellisjoe commented on okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/RemotingOkHttpCall.java:
```diff
@@ -185,81 +212,125 @@ public void onResponse(Call call, Response response) throws IOException {
+        private final Listener listener;
+
+        private RequestCompletion(
+                ScheduledExecutorService schedulingExecutor, ExecutorService executionExecutor,
```
could make this class not static and just reference the executors directly

j-baker replied:
done

ellisjoe commented on okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/RemotingOkHttpCall.java:
```diff
@@ -185,81 +212,125 @@ public void onResponse(Call call, Response response) throws IOException {
+
+        void onError(Call call, IOException exception) {
+            listener.onIgnore();
+            callback.onFailure(call, exception);
```
didn't want to inline these?

j-baker replied:
yeah, see rob's comments above

ellisjoe commented on okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/RemotingOkHttpCall.java:
```diff
@@ -140,6 +149,24 @@ private static ResponseBody buffer(MediaType mediaType, byte[] bodyBytes) {
     @Override
     public void enqueue(Callback callback) {
+        ListenableFuture<Listener> listenerFuture = limiter.acquire();
+        Futures.addCallback(listenerFuture, new FutureCallback<Listener>() {
```
when you're unable to acquire, this callback ends up getting called by another thread that is completing a some other request. that seems odd?

ellisjoe replied:
also wondering if there's some way for a request to be stuck in the enqueued state and never have it's Listener get set because there's no available slots when all in flight requests complete and then no new requests get enqueued so the queue is never drained

j-baker replied:
no, becaus ethis isn't rate limiting - it's concurrency limiting

j-baker committed:
PR comments

j-baker committed:
Tweak the concurrency limiters lib

Let's close the window as soon as a request is dropped.
Also use vegas limit by default

j-baker committed:
reset flow control test

j-baker committed:
changes

j-baker committed:
Changes

j-baker committed:
Passes the build

j-baker committed:
update lockfiles

ellisjoe commented:
have you seen the concurrency limiting in Dispatcher? https://github.com/square/okhttp/blob/master/okhttp/src/main/java/okhttp3/Dispatcher.java

It also seems like ideally we'd be able to implement the concurrency limiting at the `Call` level. You'd end up with a `ForwardingCall` that looks something like:
```
public Response execute() {
    Limiter limiter = limiters.limit(pathTemplateHeader);
    Response response = delegate.execute();
    if (response / 100 == 2) {
        limiter.success();
    } else if (....) {
        limiter.dropped();
    } else {
        limiter.ignored();
    }
}
```
which lets you avoid the async extras. Doesn't necessarily need to be exactly this, but probably worth discussing an option like this?

ellisjoe commented:
The approach mentioned above is the way Netflix does it in their GRPC implementation as well: https://github.com/Netflix/concurrency-limits/blob/master/concurrency-limits-grpc/src/main/java/com/netflix/concurrency/limits/grpc/client/ConcurrencyLimitClientInterceptor.java

j-baker committed:
Merge remote-tracking branch 'origin/develop' into jbaker/better_429_behaviour

j-baker committed:
New attempt using interceptor

j-baker committed:
more comments

ellisjoe commented on okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/ConcurrencyLimitingInterceptor.java:
```diff
@@ -0,0 +1,163 @@
+ * An interceptor for limiting the concurrency of requests to an endpoint.
+ *
+ * Requests must be tagged (before reaching this point) with a ConcurrencyLimitTag. At this point, we block on
+ * receiving a permit to run the request, and store the listener in the tag.
```
I think it would be better to just use the QosHandler directly in here rather than passing around this tag and requiring it to be set

ellisjoe commented on okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/ConcurrencyLimitingInterceptor.java:
```diff
@@ -0,0 +1,163 @@
+        return chain.proceed(chain.request());
+    }
+
+    public static Callback wrapCallback(Callback callback) {
```
don't follow why this is necessary?

ellisjoe commented on okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/ConcurrencyLimitingInterceptor.java:
```diff
@@ -0,0 +1,163 @@
+        }
+    }
+
+    private static final class ResourceDeallocator extends AsyncTimeout {
```
don't think it's worth worrying about this case. if clients aren't releasing resources properly they're going to lock things up eventually anyway. At a minimum it should be a separate change from the concurrency limiting

ellisjoe replied:
discussed more in person: going to timeout on acquiring a limit vs releasing a limit which has the added benefit of always allowing requests through after some period of time

ellisjoe commented on okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/RemotingConcurrencyLimiter.java:
```diff
@@ -0,0 +1,163 @@
+ * <li>Change code style to match Palantir baseline.</li>
+ * </ol>
+ */
+final class RemotingConcurrencyLimiter implements Limiter<Void> {
```
Not sure how to review this. Are the things listed above expected to merge upstream so we can remove this class at some point?

JacekLach commented on okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/ConcurrencyLimitingInterceptor.java:
```diff
@@ -0,0 +1,163 @@
+        ResponseBody newResponseBody =
+                ResponseBody.create(currentBody.contentType(), currentBody.contentLength(),
+                        new ReleaseConcurrencyLimitBufferedSource(currentBody.source(), tag, deallocator));
+        deallocator.timeout(1, TimeUnit.MINUTES);
```
How does this play with long streaming requests? Should I release the limiter as soon as I start streaming the data, rather than after all streaming is done?

j-baker replied:
At present the assumption is that on the server side, if you have particularly cheap requests the server should be slackening its guarantees. For this reason, I'd prefer to keep the criteria here 'is the request in flight'? because then it can match up with whatever the server is doing.

JacekLach replied:
clarified in person: in the new code the timeout only applies to the period when you are waiting to receive a limiter token; the limiter for the long-running streaming request will not time out when the request is being served.

If new requests come in and spend over a minute waiting to get a rate limiting token (because we're at max capacity for a really long time), we will circuit-break and effectively reset the rate limiting, with the assumption that this was a resource leak and not actually-at-full-capacity-unable-to-serve-a-single-request for over a minute.

j-baker committed:
some bullshit

j-baker committed:
Perfect

j-baker committed:
cleanup

j-baker committed:
Ready to go?

j-baker committed:
docs

j-baker committed:
Checkstyle

j-baker committed:
chekcstyle

JacekLach commented on okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/ConcurrencyLimitingInterceptor.java:
```diff
@@ -0,0 +1,138 @@
+        public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+            if (method.getName().equals("close") && !closed) {
+                closed = true;
+                listener.onSuccess();
```
So the closing mechanism is now implicit, in that as long as the response body is closed (either by reading it fully during json object mapping, or via `inputstream.close()` if streaming and want to finish earlier) we tag as successful?

Is it viable at some point to have more control over this? One thing that occasionally happens when streaming is your stream ends too early because an error was encountered once some data (esp. headers) was already sent. We would preferably mark those as failed.

JacekLach commented on okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/ConcurrencyLimitingInterceptor.java:
```diff
@@ -0,0 +1,138 @@
+ * 429 and 503 response codes are used for backpressure, whilst 200 -> 399 request codes are used for determining
+ * new limits and all other codes are not factored in to timings.
+ * <p>
+ * Concurrency permits are only released when the response body is closed.
```
I think this should be more visible; the last sentence on a package-private class is not exactly discoverable. Perhaps javadoc on `RetrofitClient.create` (since that's the clients we use for streaming, I believe?)

iamdanfox replied:
I've just put a section about this in the README too

j-baker committed:
Metric

iamdanfox committed:
Javadoc

iamdanfox committed:
README describes new flow control

iamdanfox committed:
Move docs -> class level javadoc

iamdanfox committed:
Rename ConcurrencyLimiters#limiter -> acquireLimiter
